### DISCUSSION TITLE
Open Browser in incognito mode

### DISCUSSION THREAD

**Discussion by phollome** [johnlindquist/kit#666] [Share]
Opens the selected browser in incognito mode and reuses your choice. Run the script with <kbd>cmd</kbd>+<kbd>Return</kbd> to reset your choice.

[Open incognito-browser in Script Kit](https://scriptkit.com/api/new?name=incognito-browser&url=https://gist.githubusercontent.com/phollome/6d09cba387703dd8d7b42137fb2e4b7f/raw/d41602bf048888554a3e51fdc448c6dd62479faa/incognito-browser.js")

```js
// Menu: Incognito Browser
// Description: Open your favorite browser in incognito/private mode
// Author: Peter Holló
// Log: false

import "@johnlindquist/kit";

const listOfBrowsers = [
  {
    name: "[F]irefox",
    description: 'Opens Firefox with the "-private-window" flag',
    value: "Firefox",
    path: "Firefox.app",
    command: 'open -na "Firefox" --args -private-window',
  },
  {
    name: "[B]rave",
    description: 'Opens Brave with the "-incognito" flag',
    value: "Brave",
    path: "Brave Browser.app",
    command: 'open -na "Brave Browser" --args -incognito',
  },
  {
    name: "[S]afari",
    description: "Runs an Apple Script which emulates key press",
    value: "Safari",
    path: "Safari.app",
    command: `
      tell application "Safari"
        activate
        tell application "System Events"
          keystroke "n" using {command down, shift down}
        end tell
      end tell`,
  },
  {
    name: "[C]hrome",
    description: 'Opens Chrome with the "-incognito" flag',
    value: "Chrome",
    path: "Google Chrome.app",
    command: 'open -na "Google Chrome" --args -incognito',
  },
  {
    name: "[O]pera",
    description: 'Opens Opera with the "-private" flag',
    value: "Opera",
    path: "Opera.app",
    command: 'open -na "Opera" --args -private',
  },
  {
    name: "[E]dge",
    description: 'Opens Edge with the "-inprivate" flag',
    value: "Edge",
    path: "Microsoft Edge.app",
    command: 'open -na "Microsoft Edge" --args -inprivate',
  },
];

async function getFavoriteBrowser(availableBrowsers) {
  const listOfAvailableBrowsers = listOfBrowsers.filter((browser) => {
    const isBrowserAvailable = availableBrowsers.some(
      (item) => item === browser.value
    );
    return isBrowserAvailable;
  });
  const selectedBrowser = await arg(
    {
      placeholder: "Select a your favorite browser",
      hint: "Press first letter for quick select",
    },
    {
      choices: listOfAvailableBrowsers,
    }
  );
  return selectedBrowser;
}

const data = await db({
  availableBrowsers: [],
});

if (data.availableBrowsers.length === 0) {
  for (let i = 0, length = listOfBrowsers.length; i < length; i++) {
    const available = await isDir(`/Applications/${listOfBrowsers[i].path}`);
    if (available) {
      data.availableBrowsers.push(listOfBrowsers[i].value);
    }
  }
  await data.write();
}

if (data.availableBrowsers.length === 0) {
  const browserNames = listOfBrowsers.map((browser) => {
    return browser.value;
  });

  await div(
    `
    <div class="p-4 font-mono text-sm">
      <p class="pb-2">
          No supported browser found!😬
      </p>
      <p>
        Supported browsers are: ${browserNames.join(", ")}.
      <p>
    </div>
    `
  );
}

const favoriteBrowser = await env("FAVORITE_BROWSER", {
  reset: flag?.cmd,
  choices: getFavoriteBrowser.bind(null, data.availableBrowsers),
});

const { command } = listOfBrowsers.find((browser) => {
  return browser.value === favoriteBrowser;
});

if (favoriteBrowser === "Safari") {
  await applescript(command);
} else {
  await exec(command);
}

```

## Initial description
To start writing scripts for this awesome tool I wanted to migrate an Alfred workflow I wrote ([alfred-incognito-browser
](https://github.com/phollome/alfred-incognito-browser)). What I appreciate on this workflow is that it enhances the application launch feature of Alfred with the option to open your browser in incognito mode. Maybe launching an application could be a top level feature instead of being a single script?

I would love get some feedback. I know there is some room for improvements (eg. the Safari script).

**Comment by johnlindquist** [comment]
Awesome, thanks for sharing!

## Modifying "Top-level" Menu

So there's technically no "top-level" menu. It's just a script mapped to a keyboard shortcut just like any other script.

So the plan, over time, is to make a few "top-level" scripts and allow people to choose/customize their own. Once the community grows more accustomed to the idea of "everything is a script", I think they'll be more ready to accept what's going on. Whether that's adding in App Launchers, File Search, etc, that can be entirely up to the user's script rather than making some complex menu with a bunch of checkboxes.

## Safari

Everything I've learned about AppleScript is from copying/pasting examples 😅. So if the internet says that we can't pass args to Safari, then I have nothing to add.

## Quick Keys

This is definitely a great use-case for quick keys. Adding a [] around a letter in the `name` of a choice will allow you to select that choice by pressing that letter 🙂

```js
const listOfBrowsers = [
  {
    name: "[F]irefox",
    path: "Firefox.app",
    command: 'open -na "Firefox" --args -private-window',
  },
  {
    name: "[B]rave",
    path: "Brave Browser.app",
    command: 'open -na "Brave Browser" --args -incognito',
  },
  {
    name: "[S]afari",
    path: "Safari.app",
    command: `
      tell application "Safari"
        activate
        tell application "System Events"
          keystroke "n" using {command down, shift down}
        end tell
      end tell`,
  },
  {
    name: "[C]hrome",
    path: "Google Chrome.app",
    command: 'open -na "Google Chrome" --args -incognito',
  },
  {
    name: "[O]pera",
    path: "Opera.app",
    command: 'open -na "Opera" --args -private',
  },
  {
    name: "[E]dge",
    path: "Microsoft Edge.app",
    command: 'open -na "Microsoft Edge" --args -inprivate',
  },
];
```

## `env`

I would probably use `await env` instead of `db` in this scenario.

```js
let functionThatReturnsFilteredChoices = // your code above

// This function will only run the first time `env` is invoked, then if will store the choice in your `~/.kenv/.env`
await env(`FAVORITE_BROWSER`, functionThatReturnsFilteredChoices)
```

**Reply by phollome** [reply]
Thanks for the feedback. I've updated the script a little bit following the hints you gave to me:
✅ Quick keys 
✅ env
😅 Safari

**Reply by phollome** [reply]
Hmm…maybe I'll remove this "favorite browser" thing. It can be frustrating to always reset the entry in the .env file, if you want to switch the browser during development or testing.

**Reply by johnlindquist** [reply]
My general thinking is that you could just "duplicate" this script (hit "right arrow" with any script selected to see the alt actions menu. Or <kbd>cmd+d</kbd>):

<img width="526" alt="CleanShot 2021-12-22 at 11 49 49@2x" src="https://user-images.githubusercontent.com/36073/147141010-c72646d3-5e3a-45d1-bb1e-d1a2a1de2517.png">

And have one script that uses "favorite" and another script that doesn't.

## `cmd` + `reset`

Another approach is to use a "command flag".

Run the script with <kbd>cmd+return</kbd> to toggle the `flag.cmd` to `true`. Then you can use that flag with the `reset` property of `env`:

```js
// Name: testing-reset-env
// cmd: Reset Env Var
import "@johnlindquist/kit"

// Prompts when unset or `reset` is true
let someEnv = await env("SOME_ENV_THAT_CHANGES_OFTEN", {
  reset: flag?.cmd,
  choices: () => ["one", "two"],
})

await div(someEnv)
```

**Reply by phollome** [reply]
Yeah, I think the <kbd>cmd+return</kbd> flag is perfect. Awesome. Thank you.
